src/rust/ide/view/graph-editor/src/component/node/output/area.rs

```diff
@@ -241,18 +265,104 @@ impl Model {
+
+    fn build_port_shapes_on_new_expression(&self) {
+        let mut port_index    = 0;
+        let mut id_crumbs_map = HashMap::new();
+        let whole_expr_type   = self.expression.borrow().whole_expr_type.clone();
+        let whole_expr_id     = self.expression.borrow().whole_expr_id;
+        let port_count        = self.port_count.get();
+        self.traverse_borrowed_expression_mut(|is_a_port,mut node,builder| {
+            let ast_id = if port_count == 0 { whole_expr_id } else { node.ast_id };
+            if let Some(id) = ast_id {
+                id_crumbs_map.insert(id,node.crumbs.clone_ref());
+            }
+
+            if DEBUG {
+                let indent  = " ".repeat(4*builder.depth);
+                let skipped = if !is_a_port { "(skip)" } else { "" };
+                println!("{}[{},{}] {} {:?} (tp: {:?}) (id: {:?})",indent,node.payload.index,
+                    node.payload.length,skipped,node.kind.variant_name(),node.tp(),node.ast_id);
+            }
+
+            if is_a_port {
+                let port   = &mut node;
+                let crumbs = port.crumbs.clone_ref();
+                let logger = &self.logger;
+                let scene  = self.scene();
+                let (port_shape,port_frp) = port.payload_mut()
+                    .init_shape(logger,scene,&self.styles,port_index,port_count);
+                let port_network = &port_frp.network;
+
+                frp::extend! { port_network
+                    self.frp.source.on_port_hover <+ port_frp.on_hover.map
+                        (f!([crumbs](t) Switch::new(crumbs.clone(),*t)));
+                    self.frp.source.on_port_press <+ port_frp.on_press.constant(crumbs.clone());
+                    port_frp.set_size_multiplier <+ self.frp.port_size_multiplier;
+                    self.frp.source.on_port_type_change <+ port_frp.tp.map(move |t|(crumbs.clone(),t.clone()));
+                }
+
+                self.ports.add_child(&port_shape);
+                port_index += 1;
+            }
+        });
+        *self.id_crumbs_map.borrow_mut() = id_crumbs_map;
+    }
+
+    fn init_definition_types(&self) {
+        let port_count        = self.port_count.get();
+        let whole_expr_type   = self.expression.borrow().whole_expr_type.clone();
+        self.traverse_borrowed_expression(|is_a_port,mut node,builder| {
+            if let Some(port_frp) = &node.payload.frp {
+                let node_tp : Option<Type> = node.tp().cloned().map(|t|t.into());
+                let node_tp = if port_count != 0 { node_tp } else {
+                    node_tp.or_else(||whole_expr_type.clone())
+                };
+                port_frp.set_definition_type.emit(node_tp);
+            }
+        })
+    }
+
+    fn set_expression(&self, new_expression:impl Into<node::Expression>) {
+        let new_expression = Expression::from(new_expression.into());
+        if DEBUG { println!("\n\n=====================\nSET EXPR: {:?}", new_expression) }
```

Comment:
Should this go to a logger?

Reply:
Nope, this is debugging tool that maybe should go one day to logger, but so far Im not sure about it. Id love to chat about it one day more, but for now Im leaving it like this. The problem with logger here is that when I enable debug mode, Im printing a very hierarhcical structure, and all the logger naes are cluttering it a lot. We should think about adding some formatters to logger, but before that happens, this way is better here.